[compiler] Improve conversion from PyTorch IR to Relay

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24 [WIP] Use PackedFunc to enable custom TVM operator registration from python
* #23 [benchmark] Add initial benchmark file
* **#22 [compiler] Improve conversion from PyTorch IR to Relay**
* #21 [testing] Improve testing to include models, image inputs and autotuned parameters
* #17 [compiler] Add None type handling with sentinel value
* #14 [compiler] Support operators with multiple outputs

bwasti committed:
[compiler] Improve conversion from PyTorch IR to Relay

bwasti commented:
This PR predominantly fixes bugs in conversion

wanchaol commented on torch_tvm/operators.cpp:
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
        auto out = tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
        return out;
      }},
+    {Symbol::fromQualString("aten::threshold_"),
```
are you getting `aten::threshold_` from the trace rather than `aten::relu`?

wanchaol replied:
`aten::threshold` is too generic to pass to tvm, I think we should only consider `aten::relu`

bwasti replied:
I'm getting threshold_ from the trace, not relu

wanchaol commented on torch_tvm/version.py:
```diff
@@ -0,0 +1,9 @@
+# This file is generated by setup.py. DO NOT EDIT!
```
what's this used for? can we add it to gitignore instead?

bwasti replied:
ah this shouldn't even be in this commit

wanchaol commented on torch_tvm/operators.cpp:
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
+       auto d = relayToConstant<float>(inputs[1]);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for threshold 0, got", d);
+       AT_CHECK(d > -1e-7, "aten::threshold_ only supported for threshold 0, got", d);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for threshold 0, got", d);
```
it looks like there are 2 duplicate checks here

wanchaol commented on torch_tvm/operators.cpp:
```diff
@@ -186,6 +217,25 @@ RegisterTVMOperator reg({
+       d = relayToConstant<float>(inputs[2]);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for value 0, got", d);
+       AT_CHECK(d > -1e-7, "aten::threshold_ only supported for value 0, got", d);
+       AT_CHECK(d <  1e-7, "aten::threshold_ only supported for value 0, got", d);
```
same here

wanchaol commented on torch_tvm/compiler.cpp:
```diff
@@ -23,24 +28,33 @@ tvm::relay::Var TVMCompiler::convertToRelay(const Value* val) {
 }
 
 tvm::relay::Expr TVMCompiler::convertToRelay(const IValue& val) {
+  // All doubles are converted to floats
```
looks like we can convert to `float64` for double, can you add a comment here why the conversion to `float32` is necessary?

bwasti replied:
it's not really necessary, I'll swap to fp64

wanchaol commented on torch_tvm/compiler.cpp:
```diff
@@ -69,14 +85,15 @@ tvm::relay::Expr TVMCompiler::convertToRelay(const IValue& val) {
 
 tvm::relay::Function TVMCompiler::convertToRelay(
-    std::shared_ptr<Graph> subgraph) {
+    std::shared_ptr<Graph> subgraph, std::vector<Value*>* input_values) {
```
just take a reference of the vector as argument?

bwasti replied:
input_values is being written to, so I'm passing it as a pointer

wanchaol commented on torch_tvm/compiler.cpp:
```diff
@@ -199,11 +224,21 @@ void TVMCompiler::run(Stack& stack) {
+      value_to_ivalue[value] = optional_ivalue.value();
+    }
+    auto ivalue = value_to_ivalue.at(input_values_[i]);
+    auto tensor = ivalue.toTensor().to(caffe2::TypeMeta::Make<float>());
```
I think we can do something like `ivalue.toTensor().to(at::kFloat)`

bwasti committed:
Update on "[compiler] Improve conversion from PyTorch IR to Relay"

[compiler] Improve conversion from PyTorch IR to Relay

gh-metadata: pytorch tvm 22 gh/bwasti/10/head

bwasti committed:
Update on "[compiler] Improve conversion from PyTorch IR to Relay"

[compiler] Improve conversion from PyTorch IR to Relay

gh-metadata: pytorch tvm 22 gh/bwasti/10/head

bwasti committed:
Update on "[compiler] Improve conversion from PyTorch IR to Relay"

[compiler] Improve conversion from PyTorch IR to Relay

gh-metadata: pytorch tvm 22 gh/bwasti/10/head

bwasti committed:
Update on "[compiler] Improve conversion from PyTorch IR to Relay"

[compiler] Improve conversion from PyTorch IR to Relay

gh-metadata: pytorch tvm 22 gh/bwasti/10/head
